tests/unit/lms/services/lis_result_sourcedid_test.py

```diff
@@ -1,187 +1,191 @@
-            lis_result_sourcedid_info_2, h_user2, lti_user
-        )
+        pyramid_request.POST["lis_result_sourcedid"] = "something different"
 
-        assert lis_result_sourcedid_updated is lis_result_sourcedid
-        assert (
-            lis_result_sourcedid_updated.lis_result_sourcedid == "something different"
-        )
-        assert lis_result_sourcedid.h_display_name == "Someone Else"
+        svc.upsert_from_request(pyramid_request, h_user, lti_user)
 
+        result = self.get_inserted_row(db_session)
+        assert result == Any.instance_of(LISResultSourcedId)
+        assert result.lis_result_sourcedid == "something different"
+        assert result.h_display_name == "Someone Else"
 
-def _add_lis_result_sourcedid(lis_result_sourcedid_info, h_user, lti_user, db_session):
-    lis_result_sourcedid_ = LISResultSourcedId()
-    lis_result_sourcedid_.lis_result_sourcedid = (
-        lis_result_sourcedid_info.lis_result_sourcedid
-    )
-    lis_result_sourcedid_.lis_outcome_service_url = (
-        lis_result_sourcedid_info.lis_outcome_service_url
-    )
-    lis_result_sourcedid_.context_id = lis_result_sourcedid_info.context_id
-    lis_result_sourcedid_.resource_link_id = lis_result_sourcedid_info.resource_link_id
-    lis_result_sourcedid_.tool_consumer_info_product_family_code = (
-        lis_result_sourcedid_info.tool_consumer_info_product_family_code
+    @pytest.mark.parametrize(
+        "param",
+        (
+            "lis_result_sourcedid",
+            "lis_outcome_service_url",
+            "context_id",
+            "resource_link_id",
+        ),
     )
-    lis_result_sourcedid_.h_username = h_user.username
-    lis_result_sourcedid_.h_display_name = h_user.display_name
-    lis_result_sourcedid_.oauth_consumer_key = lti_user.oauth_consumer_key
-    lis_result_sourcedid_.user_id = lti_user.user_id
+    def test_it_does_nothing_with_required_parameter_missing(
+        self, svc, pyramid_request, h_user, lti_user, db_session, param
+    ):
+        del pyramid_request.POST[param]
 
-    db_session.add(lis_result_sourcedid_)
-    return lis_result_sourcedid_
+        svc.upsert_from_request(pyramid_request, h_user, lti_user)
+
+        assert self.get_inserted_row(db_session) is None
+
+    @pytest.mark.parametrize("param", ("tool_consumer_info_product_family_code",))
+    def test_it_works_fine_with_optional_parameter_missing(
+        self, svc, pyramid_request, h_user, lti_user, db_session, param
+    ):
+        del pyramid_request.POST[param]
+
+        svc.upsert_from_request(pyramid_request, h_user, lti_user)
+
+        assert self.get_inserted_row(db_session) == Any.instance_of(LISResultSourcedId)
+
+    @classmethod
+    def model_as_dict(cls, model):
+        return {col.key: getattr(model, col.key) for col in model.iter_columns()}
+
+    @classmethod
+    def get_inserted_row(cls, db_session):
+        return db_session.query(LISResultSourcedId).one_or_none()
```

Comment:
Err... when will that get evaluated?

Reply:
## Signatures are already long

I like it... but:

```
def test_it_updates_existing_record_if_matching_exists(
        self,
        svc,
        pyramid_request,
        h_user,
        lti_user,
        lis_result_sourcedid,
        db_session,
        lti_params,
    ):
```
I'm not sure that really needs another parameter just so we don't call self. A few of the `lis` related tests have a similar problem where the method signatures are so long it's hard to read the tests.

## It's a bit weird

I suspect this might also be a case of the principal of least surprise. This is not generally how the fixtures are used. On the whole our tests are written so that fixtures are things being tested or setup, rather than being helper functions.

The first time I read that I did not spot the last set of brackets were missing in the function, and I didn't understand how it worked.

## Enclosing db_sessions

This SO question has two answers to the same question:
https://stackoverflow.com/questions/45923876/pass-a-fixture-to-a-helper-function-in-pytest

a) The same solution proposed here
b) Observing that we really want `db_session` to be accessible in `get_inserted_row`, maybe there is a missing wrapper around both.

I've tried this:

```
    @pytest.fixture
    def db_session(self, db_session):
        db_session.get_last_inserted = lambda: db_session.query(LISResultSourcedId).one_or_none()
        return db_session
# ...

    assert db_session.get_last_inserted() == Any.instance_of(LISResultSourcedId)
```

So this works, it's very short and is "cute". I'm not sure it's the good kind of cute though. It requires good knowledge of `pytest` and introduces more magical state that appears from nowhere.

I wonder if it's good to remember that 100% of people who are very familiar with `pytest` are familiar with python, but the reverse isn't true. People who understand this will understand `@classmethod` but not vice versa.

Reply:
`db_session.get_last_inserted()` seems good to me. I would niggle that the lambda is unnecessary (the attribute is already called **get**_last_inserted so it's explicitly a method). But it's fine.

Some comments by way of discussion rather than PR review/not requesting further changes:

## Long test method param lists

I agree that really long parameter lists is a code smell but I think that's telling us something about the code under test, not about the tests themselves. Why does it take so many fixtures to test this service? I'm not suggesting we try to address any code design issue in this PR though. But just to point out: this is a deliberate aspect of the design of pytest fixtures and why they're better than traditional test class setup and teardown methods and `self`: each test explicitly declares, in its argument list, what its dependencies are. It's the "explicit" in the ["explicit, modular, scalable"](https://docs.pytest.org/en/latest/fixture.html#fixture) motto of pytest fixtures.

That doesn't mean that long argument lists are okay. But I think moving arguments onto `self` is just hiding the problem (of too many dependencies) not fixing it, while making the test methods less explicit and localized.

## Required pytest knowledge

I had a similar reaction to pytest when I joined Hypothesis. I was used to nosetests from my previous job and seeing pytest fixtures being used led me to ask "Why not just?" (write a test helper method, use `self` and `setUp()`, etc). Nowadays I'd argue the same thing that Nick argued at that time: pytest is a _really_ great test framework, the best, and taking the time to learn it is really going to improve our tests. And nowadays we do have [an extensive tutorial on it](https://seanh.cc/post/python-unit-testing/) after all (see particularly the two posts on fixtures), plus pytest's own docs.

Both pytest-based solutions (my `get_inserted_row()` fixture and your `db_session.get_last_inserted()` method) seem fine to me. They don't seem to me to be too surprising or require too much pytest knowledge. For example the `get_inserted_row()` fixture:

1. Is a pytest fixture. So you have to basically know what those are
2. Depends on the `db_session` fixture. So you have to know about pytest fixture composition
3. Returns a method. So you have to know that methods can be used as return values and arguments in Python, and that that's perfectly compatible with pytest fixtures (why wouldn't it be?).

   You also have to spot that this fixture, unlike most, returns a method, but that's in the fixture name (`get_inserted_row`, a fixture that returned the row itself rather than a method for getting the row would always be called `inserted_row` instead), and can also be mentioned in a code comment or fixture docstring.

   Although they're not as common as non-callable fixtures we do in fact use callable fixtures elsewhere in our tests already and the fixture name is always a verb (or begins with a verb) rather than a noun.

There's certainly some pytest knowledge required there. But it doesn't feel excessive to me.